code/datums/components/echolocation.dm

```diff
@@ -0,0 +1,204 @@
+	var/image_expiry_time = 1.5 SECONDS
+	/// Time for the image to fade in.
+	var/fade_in_time = 0.5 SECONDS
+	/// Time for the image to fade out and delete itself.
+	var/fade_out_time = 0.5 SECONDS
+	/// Are images static? If yes, spawns them on the turf and makes them not change location. Otherwise they change location and pixel shift with the original.
+	var/images_are_static = TRUE
+	/// With mobs that have this echo group in their echolocation receiver trait, we share echo images, defaults to quirk
+	var/echo_group = null
+	/// Color applied over the client
+	var/client_color = null
+	/// Associative list of world.time when created to a list of the images.
+	var/list/images = list()
+	/// Associative list of world.time when created to a list of receivers.
+	var/list/receivers = list()
+	/// All the saved appearances, keyed by icon-icon_state.
+	var/static/list/saved_appearances = list()
+	/// Typecache of all the allowed paths to render.
+	var/static/list/allowed_paths
+	/// Typecache of turfs that are dangerous, to give them a special icon.
+	var/static/list/danger_turfs
+	/// A matrix that turns everything except #ffffff into pure blackness, used for our images (the outlines are #ffffff).
+	var/static/list/black_white_matrix = list(85, 85, 85, 0, 85, 85, 85, 0, 85, 85, 85, 0, 0, 0, 0, 1, -254, -254, -254, 0)
+	/// A matrix that turns everything into pure white.
+	var/static/list/white_matrix = list(255, 255, 255, 0, 255, 255, 255, 0, 255, 255, 255, 0, 0, 0, 0, 1, 0, -0, 0, 0)
+
+	/// Cooldown for the echolocation.
+	COOLDOWN_DECLARE(cooldown_last)
+
+/datum/component/echolocation/Initialize(echo_range, cooldown_time, image_expiry_time, fade_in_time, fade_out_time, images_are_static, blocking_trait, echo_group, echo_icon, color_path)
+	. = ..()
+	var/mob/living/echolocator = parent
+	if(!istype(echolocator))
+		return COMPONENT_INCOMPATIBLE
+	if(!danger_turfs)
+		danger_turfs = typecacheof(list(/turf/open/chasm, /turf/open/lava, /turf/open/floor/fakepit))
+	if(!allowed_paths)
+		allowed_paths = typecacheof(list(/turf/closed, /obj, /mob/living)) + danger_turfs - typecacheof(/obj/effect/decal)
+	if(!isnull(echo_range))
+		src.echo_range = default_echo_range
+	if(!isnull(cooldown_time))
+		src.cooldown_time = cooldown_time
+	if(!isnull(image_expiry_time))
+		src.image_expiry_time = image_expiry_time
+	if(!isnull(fade_in_time))
+		src.fade_in_time = fade_in_time
+	if(!isnull(fade_out_time))
+		src.fade_out_time = fade_out_time
+	if(!isnull(images_are_static))
+		src.images_are_static = images_are_static
+	if(ispath(color_path))
+		client_color = echolocator.add_client_colour(color_path)
+	else
+		client_color = echolocator.add_client_colour(/datum/client_colour/echolocate)
+
+	src.echo_group = echo_group || REF(src)
+	echolocator.add_traits(list(TRAIT_ECHOLOCATION_RECEIVER, TRAIT_NIGHT_VISION), echo_group) //so they see all the tiles they echolocated, even if they are in the dark
+	echolocator.become_blind(TRAIT_BLIND_ECHO)
```

Comment:
I assume you meant to use ``become_blind_echolocation`` here instead?

Though why do we do that instead of just using ``become_blind``'s source?

Reply:
no, we want to add the blindness trait with echolocation because blind_echolocation doesnt cover all the bases that blindness does

Reply:
but i think it is typed wrong, ill look over it in a bit